### PR TITLE
[runtime env] Add temporary URI reference to prevent URI deletion before job starts

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
Packages are uploaded to the GCS for `runtime_env`.  These packages are garbage collected when their refcount becomes zero.

The problem is the reference doesn't get incremented until the job starts, which happens after the package is uploaded.  It's possible for the package's refcount to go to zero in between the upload and when the job starts, causing the package to be deleted before it's needed by the job.  It's likely the cause of https://github.com/ray-project/ray/issues/23423.

We can't just increment the refcount at the time of upload, because if the script is killed before the job is started (e.g. via Ctrl-C) then the reference will never be decremented and the package will never be deleted.

The solution in this PR is to increment the refcount at the time of upload, but automatically decrement after a configurable timeout (default 30s).  This should be enough time for the job to start.  When the job starts, it increments the refcount as usual and decrements it when the job finishes or is killed.

```
upload_package
> refcount: 1
AddJob
> refcount:2
--- 30s pass
> refcount: 1
job ends (could happen before 30s pass)
> refcount: 0
```

TODO before merge:
- [x] Call the RPC in the appropriate place so that the reference gets incremented in both cases: when the package already exists, and when it doesn't exist.  We also need to call the RPC in the Ray Job Submission codepath.
- [x] Add test for ray.init()
- [x] Add test for job submission
- [x] Tests passing

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by edoakes — COMMENTED** [review]
@architkulkarni doesn't the increment need to be made atomic with the call to check if the URI exists?

**Comment by architkulkarni** [comment]
> @architkulkarni doesn't the increment need to be made atomic with the call to check if the URI exists?

Hmm, let me see if I understand correctly... Currently the approach is simple, we always do the increment.  So when `upload_package_if_needed` returns, it guarantees that both (1) the package exists and (2) it has a temporary reference so it won't get deleted until at least 10 minutes.

Because it's not atomic, it could happen that the program crashes after the call to `add_temporary_reference` but before we check if the package exists and upload it if needed.  This causes the bad outcome that we have a temporary reference for a package which either doesn't exist in the GCS, or that does exist but isn't needed by any job.  But the impact of this seems small: We'll either try to kv_delete a package which doesn't exist (which I think is just a no-op), or we'll delete a package 10 minutes later than we could have.

From that perspective, it seems like it would be a minor optimization for a rare failure case so I'm not sure it's worth the added complexity.

How would we make it atomic?  Even if we moved this part to C++ and used internal functions, couldn't the same thing happen (something fails in between incrementing the reference and checking if the package exists, depending on which order we do these two operations)?

**Comment by edoakes** [comment]
Resolved offline, agreed with the approach currently in the PR

File: dashboard/modules/dashboard_sdk.py [hunk 1]
```diff
@@ -311,6 +317,8 @@ def _upload_package_if_needed(
         else:
             package_uri = get_uri_for_directory(package_path, excludes=excludes)
 
+        self._increment_temporary_reference(package_uri)
```
Comment (edoakes) [inline dashboard/modules/dashboard_sdk.py:320]: please add a comment about why this is necessary, point to the GH issue

File: dashboard/modules/job/job_head.py [hunk 2]
```diff
@@ -103,6 +107,23 @@ async def upload_package(self, req: Request):
...
+        )
+        logger.debug(f"Adding temporary reference to package {package_uri}.")
+        try:
+            add_temporary_uri_reference(package_uri)
+        except Exception:
+            return Response(
+                text=traceback.format_exc(),
+                status=aiohttp.web.HTTPInternalServerError.status_code,
+            )
+        return Response(status=aiohttp.web.HTTPOk.status_code)
```
Comment (edoakes) [inline dashboard/modules/job/job_head.py:125]: I don't think we need to add a separate API endpoint for this (we should ideally treat API endpoints are ~~ public APIs), I'd prefer if we just do this in the existing API endpoint we use to check if the ref exists.

File: python/ray/_private/runtime_env/packaging.py [hunk 3]
```diff
@@ -237,6 +238,25 @@ def match(p: Path):
         return None
 
 
+def add_temporary_uri_reference(uri: str) -> None:
```
Comment (edoakes) [inline python/ray/_private/runtime_env/packaging.py:241]: link to GH issue

File: python/ray/_private/runtime_env/packaging.py [hunk 4]
```diff
@@ -237,6 +238,25 @@ def match(p: Path):
...
+    Packages is uploaded to GCS in order to be downloaded by
+    a runtime env plugin (e.g. working_dir, py_modules) after the job starts. Add a
+    temporary reference to the package in the GCS to prevent it from being deleted
+    before the job starts. If this reference didn't have an expiration, then if the
+    script exited (e.g. via Ctrl-C) before the job started, the reference would never
+    be removed, so the package would never be deleted from the GCS.
+    """
+
+    TEMPORARY_REFERENCE_EXPIRATION_S = int(
+        os.environ.get("RAY_runtime_env_temporary_reference_expiration_s", 60 * 10)
```
Comment (edoakes) [inline python/ray/_private/runtime_env/packaging.py:253]: Put the env var name in `ray_constants.py`

File: src/ray/gcs/gcs_server/runtime_env_handler.cc [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+  runtime_env_manager_.AddURIReference(hex_id, request.uri());
+  GCS_RPC_SEND_REPLY(send_reply_callback, reply, Status::OK());
+  delay_executor_(
+      [this, hex_id, request] {
+        runtime_env_manager_.RemoveURIReference(hex_id);
+        RAY_LOG(DEBUG) << "Removed temporary URI reference for ID " << hex_id
+                       << "with URI:" << request.uri();
+      },
+      request.expiration_s() * 1000);
```
Comment (edoakes) [inline src/ray/gcs/gcs_server/runtime_env_handler.cc:38]: nit: like to add comment with param name like:
```suggestion
      /* expiration_ms= */ request.expiration_s() * 1000);
```

File: src/ray/protobuf/gcs_service.proto [hunk 6]
```diff
@@ -501,6 +501,20 @@ service InternalKVGcsService {
...
+}
+
+message AddTemporaryURIReferenceReply {
+  GcsStatus status = 1;
+}
+
+service RuntimeEnvGcsService {
+  rpc AddTemporaryURIReference(AddTemporaryURIReferenceRequest)
+      returns (AddTemporaryURIReferenceReply);
+}
```
Comment (edoakes) [inline src/ray/protobuf/gcs_service.proto:516]: can you add a comment about what this service is for?

File: dashboard/modules/job/job_head.py (hunk 2 above)
Comment (architkulkarni) [reply dashboard/modules/job/job_head.py:125]: Good point, that's cleaner.

**Comment by architkulkarni** [comment]
Thanks! I've addressed the comments.

**Review by ckw017 — COMMENTED** [review]
`ray_client.proto` part looks good, can you bump `CURRENT_PROTOCOL_VERSION` in `python/ray/util/client/__init__.py`

**Comment by architkulkarni** [comment]
Yeah, it's the second case.  It makes sense that you already considered this when implementing it, but it might have gotten inadvertently moved around during some refactoring afterwards.  Right now it's called here https://github.com/ray-project/ray/blob/2627c7b5bca27e6ff15d913c4e867de573f4442e/python/ray/worker.py#L1575 and here https://github.com/ray-project/ray/blob/2627c7b5bca27e6ff15d913c4e867de573f4442e/python/ray/util/client/worker.py#L792 and here (for Ray Job Submission): https://github.com/ray-project/ray/blob/2627c7b5bca27e6ff15d913c4e867de573f4442e/dashboard/modules/job/sdk.py#L123

I think in all these cases it's before the GCS is aware there's a job, but I think the job submission case is the most extreme, since we upload the package in one HTTP request and then start the job in a second HTTP request.

**Comment by iycheng** [comment]
Got it. I think I failed to take care of case 2 at that time. I think introducing timing potentially is still not working in some cases, for example, if the job starts is too slow, like 20mins because either the machine is slow or the pkg to install is too big, we still have some chance case 2 still can happen. But with this, it's still configurable. I'm not strong against this approach. The component missing here is some coordinator to manage these URIs, I thought we were moving them into agents.

My overall feeling is that it exposes too many lower level APIs related to the implementations. Do you think it could be better for the reader to understand if we abstract this feature to something like `PinPackageURI(a_list_of_uris, duration)`?

Also, I feel for ray client, it seems that we missing some functionality to expose GCS APIs to the client, with which it'll be cleaner to implement the feature.

**Comment by edoakes** [comment]
@iycheng you're right that it's still possible you could run into the same edge condition, but this should cover the vast majority of cases and we should not recommend dynamically uploading the working_dir for cases that have high stability requirements anyways (like production job scheduling). We discussed some more "correct" solutions but they required significant refactoring so we landed on this approach to unblock the users who are running into the issue for now.

**Comment by iycheng** [comment]
I'm OK with this. Could we update the add_temporary_uris_refs to pin_runtime_env_uris([uris], duration)  which I feel is easier to understand the purpose of the operation.

**Review by ckw017 — APPROVED** [review]
`ray_client.proto` part is good

**Comment by edoakes** [comment]
@iycheng sounds good. @architkulkarni is on vacation now so I will take this over and update the naming as you suggest.

File: python/ray/ray_constants.py [hunk 7]
```diff
@@ -67,6 +67,11 @@ def env_bool(key, default):
 RAY_ADDRESS_ENVIRONMENT_VARIABLE = "RAY_ADDRESS"
 RAY_NAMESPACE_ENVIRONMENT_VARIABLE = "RAY_NAMESPACE"
 RAY_RUNTIME_ENV_ENVIRONMENT_VARIABLE = "RAY_RUNTIME_ENV"
+RAY_RUNTIME_ENV_TEMP_REF_EXPIRATION_S_ENV_VAR = (
```
Comment (iycheng) [inline python/ray/ray_constants.py:70]: Update the name?

**Comment by edoakes** [comment]
ping @rkooo567 review

File: python/ray/ray_constants.py (hunk 7 above)
Comment (edoakes) [reply python/ray/ray_constants.py]: done

File: python/ray/util/client/server/proxier.py [hunk 8]
```diff
@@ -520,6 +520,26 @@ def KVExists(self, request, context=None) -> ray_client_pb2.KVExistsResponse:
...
+        Otherwise, we proxy the call to the downstream server as usual.
+        """
+        if self._has_channel_for_request(context):
+            return self._call_inner_function(request, context, "PinRuntimeEnvURI")
+
+        with disable_client_hook():
+            ray.experimental.internal_kv._pin_runtime_env_uri(
+                request.uri, expiration_s=request.expiration_s
+            )
+        return ray_client_pb2.ClientAddTemporaryURIReferenceResponse()
```
Comment (iycheng) [inline python/ray/util/client/server/proxier.py:541]: Sorry, I missed this part. Let's also update this?

**Review by fishbone — APPROVED** [review]
LGTM. Overall I feel it's not a big change, but somehow e2e, the PR is big.  :(

**Comment by edoakes** [comment]
@iycheng yeah I agree, it's almost all boilerplate for various grpc services...

File: python/ray/util/client/server/proxier.py (hunk 8 above)
Comment (edoakes) [reply python/ray/util/client/server/proxier.py:541]: good catch

File: src/ray/gcs/gcs_server/runtime_env_handler.cc [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  runtime_env_manager_.AddURIReference(hex_id, request.uri());
+  GCS_RPC_SEND_REPLY(send_reply_callback, reply, Status::OK());
+  delay_executor_(
+      [this, hex_id, request] {
+        runtime_env_manager_.RemoveURIReference(hex_id);
+        RAY_LOG(DEBUG) << "Removed temporary URI reference for ID " << hex_id
+                       << "with URI:" << request.uri();
+      },
+      /* expiration_ms= */ request.expiration_s() * 1000);
+  return;
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/runtime_env_handler.cc:38]:
```suggestion

```

**Review by rkooo567 — APPROVED** [review]
LGTM. Some minor comments to improve error message!

File: python/ray/_private/gcs_utils.py [hunk 10]
```diff
@@ -279,6 +282,18 @@ def internal_kv_keys(
                 f"due to error {reply.status.message}"
             )
 
+    @_auto_reconnect
+    def pin_runtime_env_uri(self, uri: str, expiration_s: int) -> None:
+        req = gcs_service_pb2.PinRuntimeEnvURIRequest(
+            uri=uri, expiration_s=expiration_s
+        )
+        reply = self._runtime_env_stub.PinRuntimeEnvURIRequest(req)
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/_private/gcs_utils.py:290]: Following the error handling that's done for the rest of the GCS client. I think if we want to handle specific gRPC exceptions we should do it throughout all of the RPCs.

File: python/ray/_private/runtime_env/packaging.py [hunk 11]
```diff
@@ -237,6 +242,36 @@ def match(p: Path):
...
+    This function adds a temporary reference to the package in the GCS to prevent
+    it from being deleted before the job starts. (See #23423 for the bug where
+    this happened.)
+
+    If this reference didn't have an expiration, then if the script exited
+    (e.g. via Ctrl-C) before the job started, the reference would never be
+    removed, so the package would never be deleted.
+    """
+
+    if expiration_s is None:
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/_private/runtime_env/packaging.py:263]: yeah will do

File: python/ray/_private/gcs_utils.py [hunk 12]
```diff
@@ -279,6 +282,18 @@ def internal_kv_keys(
                 f"due to error {reply.status.message}"
             )
 
+    @_auto_reconnect
+    def pin_runtime_env_uri(self, uri: str, expiration_s: int) -> None:
+        req = gcs_service_pb2.PinRuntimeEnvURIRequest(
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/_private/gcs_utils.py:287]: yup it's just one call on startup

[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/_private/gcs_utils.py:287]: I can leave a comment

File: python/ray/_private/gcs_utils.py [hunk 13]
```diff
@@ -279,6 +282,18 @@ def internal_kv_keys(
...
+            uri=uri, expiration_s=expiration_s
+        )
+        reply = self._runtime_env_stub.PinRuntimeEnvURIRequest(req)
+        if reply.status.code != GcsCode.OK:
+            raise RuntimeError(
+                f"Failed to pin URI reference for {uri} "
+                f"due to error {reply.status.message}."
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/_private/gcs_utils.py:294]: done

File: python/ray/_private/gcs_utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/_private/gcs_utils.py:291]: nvm, i fixed it, but we should probably do this for the others too

**Pull request merged**